File: 06-observability/03-correlation-ids/nodejs/lose_the_context.js
/*
 * Layer 6 Topic 3 - Losing trace context at a Node.js concurrency boundary.
 *
 * What this demonstrates
 * ----------------------
 * Node's answer to in-process context is `AsyncLocalStorage`, built on
 * async_hooks. It is the most automatic of the six: promises, timers, I/O
 * callbacks and `await` all keep the store, because the runtime tracks the
 * async resource that scheduled the continuation.
 *
 * It loses context in exactly one shape, and this file is that shape: a
 * library that manages its OWN callback queue and invokes your callback from
 * a different async resource than the one that registered it. Old connection
 * pools, hand-rolled event emitters and batching clients all do this. The
 * symptom in production is that one integration silently starts a new trace
 * while every other part of the service is fine -- so it reads as "that
 * library is weird" rather than as a propagation bug.
 *
 * The fix is `AsyncResource.bind`, which captures the current async context at
 * bind time and restores it when the function is later called from anywhere.
 * That is what a well-behaved library does for you, and what an old one does
 * not.
 *
 * No OpenTelemetry SDK is imported -- none is installed on this machine. The
 * span, the W3C `traceparent` codec and the leaky pool are all here in the
 * file, because the failure belongs to the runtime and not to an SDK.
 *
 * What to look for in the output
 * ------------------------------
 * Four blocks in one shape:
 *
 *   caller trace_id   <id>
 *   callee trace_id   <id or "none">   naive
 *   callee trace_id   <id>             propagated
 *   verdict           lost | preserved
 *
 * The setTimeout/promise block is the control: preserved, with no work from
 * you. The pool block is the same program losing the same store, because the
 * callback is fired from a timer that was registered outside any request.
 * Then read the last section: the ONE thread running every request means a
 * lost store is not a per-request accident, it is whatever the event loop
 * happened to be doing at the time.
 */

'use strict';

const { AsyncLocalStorage, AsyncResource } = require('node:async_hooks');
const crypto = require('node:crypto');

// ---------------------------------------------------------------------------
// A minimal span + W3C traceparent codec: the entire cross-process half.
// ---------------------------------------------------------------------------

function makeSpan(name, traceId) {
  return {
    name,
    traceId: traceId || crypto.randomBytes(16).toString('hex'),
    spanId: crypto.randomBytes(8).toString('hex'),
    sampled: true,
  };
}

function traceparentOf(span) {
  // version-traceid-spanid-flags
  return `00-${span.traceId}-${span.spanId}-${span.sampled ? '01' : '00'}`;
}

function spanFromTraceparent(header, name) {
  const [version, traceId, parentId, flags] = header.split('-');
  if (version !== '00' || traceId.length !== 32 || parentId.length !== 16) {
    throw new Error(`malformed traceparent: ${header}`);
  }
  const span = makeSpan(name, traceId);
  span.sampled = (parseInt(flags, 16) & 1) === 1;
  return span;
}

const als = new AsyncLocalStorage();
const currentTraceId = () => (als.getStore() ? als.getStore().traceId : 'none');

// ---------------------------------------------------------------------------
// Structured logging: read the store PER RECORD. A logger that reads it once
// at module load stamps every line with the same id, which looks like it works
// right up until you query by one.
// ---------------------------------------------------------------------------

const capturedLogs = [];
function log(msg) {
  capturedLogs.push({ msg, trace_id: als.getStore() ? als.getStore().traceId : '' });
}

function report(boundary, caller, naive, propagated, note) {
  const verdict = naive === caller ? 'preserved' : 'lost';
  console.log(`boundary          ${boundary}`);
  console.log(`caller trace_id   ${caller}`);
  console.log(`callee trace_id   ${naive.padEnd(32)} naive`);
  console.log(`callee trace_id   ${propagated.padEnd(32)} propagated`);
  console.log(`verdict           ${verdict}${note ? `   (${note})` : ''}`);
  console.log();
  return verdict;
}

// ---------------------------------------------------------------------------
// Boundary 1: await / setTimeout / promise chain -- the control.
// ---------------------------------------------------------------------------

async function boundaryAwait() {
  const span = makeSpan('GET /orders');
  return als.run(span, async () => {
    await new Promise((resolve) => setTimeout(resolve, 5));
    const afterTimer = currentTraceId();
    const afterPromise = await Promise.resolve().then(() => currentTraceId());
    const observed = afterTimer === afterPromise ? afterTimer : 'DISAGREE';
    return report(
      'await / setTimeout / .then',
      span.traceId,
      observed,
      observed,
      'async_hooks follows the continuation; nothing to fix',
    );
  });
}

// ---------------------------------------------------------------------------
// Boundary 2: a library with its own callback queue. This is the leaky shape.
//
// LeakyPool stores your callback in an array and fires it later from a timer
// that IT registered, at module scope, before your request existed. The
// callback runs inside the timer's async context, not yours.
// ---------------------------------------------------------------------------

class LeakyPool {
  constructor() {
    this.pending = [];
    // Registered once, outside any request. This is the async resource that
    // will end up invoking every callback.
    this.drainer = setInterval(() => this.drain(), 5);
  }

  query(sql, cb) {
    this.pending.push({ sql, cb });
  }

  drain() {
    const batch = this.pending.splice(0, this.pending.length);
    for (const { sql, cb } of batch) cb(null, { sql, rows: 1 });
  }

  close() {
    clearInterval(this.drainer);
  }
}

function boundaryPool(pool) {
  const span = makeSpan('GET /orders');
  return als.run(span, () => new Promise((resolve) => {
    let naive = null;
    let propagated = null;

    // Naive: hand the pool a bare closure.
    pool.query('SELECT 1', () => {
      log('pool callback (naive)');
      naive = currentTraceId();
      maybeDone();
    });

    // Propagated: bind the callback to the CURRENT async context first. This
    // is precisely what a modern client library does on your behalf.
    pool.query('SELECT 1', AsyncResource.bind(() => {
      log('pool callback (bound)');
      propagated = currentTraceId();
      maybeDone();
    }));

    function maybeDone() {
      if (naive === null || propagated === null) return;
      resolve(report(
        'pool with its own callback queue',
        span.traceId,
        naive,
        propagated,
        'fix = AsyncResource.bind(cb) at the point you hand it over',
      ));
    }
  }));
}

// ---------------------------------------------------------------------------
// Boundary 3: a queue. No async resource is involved at all -- the job is a
// row in a table. Only the message body can carry the context.
// ---------------------------------------------------------------------------

function boundaryQueue() {
  const span = makeSpan('POST /orders');
  return als.run(span, () => {
    const jobs = [
      { id: 'naive', customer: 'cust-0042' },
      { id: 'propagated', customer: 'cust-0042', traceparent: traceparentOf(span) },
    ];

    const seen = {};
    for (const job of jobs) {
      // The consumer is a different process in the lab. Here it is a function
      // called with no ambient context, which is the same thing.
      const consumed = als.run(undefined, () => {
        if (job.traceparent) {
          return als.run(spanFromTraceparent(job.traceparent, 'job'), () => {
            log(`processing job ${job.id}`);
            return currentTraceId();
          });
        }
        log(`processing job ${job.id}`);
        return currentTraceId();
      });
      seen[job.id] = consumed;
    }

    return report(
      'Postgres-backed queue',
      span.traceId,
      seen.naive,
      seen.propagated,
      'the transport carries no headers; put traceparent in the body',
    );
  });
}

// ---------------------------------------------------------------------------
// Boundary 4: the outbound HTTP call -- the easy half, made concrete.
// ---------------------------------------------------------------------------

function boundaryHttp() {
  const span = makeSpan('GET /orders');
  const header = traceparentOf(span);
  const downstream = spanFromTraceparent(header, 'GET /price');
  console.log('boundary          HTTP request to pricing');
  console.log(`caller trace_id   ${span.traceId}`);
  console.log(`traceparent sent  ${header}`);
  console.log(`callee trace_id   ${downstream.traceId.padEnd(32)} parsed from the header`);
  console.log('verdict           preserved   (this is what being a W3C standard buys)');
  console.log();
  return 'preserved';
}

// ---------------------------------------------------------------------------
// The Node-specific closing point: interleaving. Two requests are in flight on
// ONE thread. With the store, each callback still knows which request it
// belongs to; without it, "the current request" is a global that the last
// arrival overwrote.
// ---------------------------------------------------------------------------

let ambientGlobal = null; // the thing people write instead of AsyncLocalStorage

async function interleaved() {
  const results = [];

  async function handle(name, delayMs) {
    const span = makeSpan(name);
    ambientGlobal = span; // the naive "current request" variable
    return als.run(span, async () => {
      await new Promise((r) => setTimeout(r, delayMs));
      results.push({
        request: name,
        withStore: currentTraceId(),
        withGlobal: ambientGlobal.traceId,
        correct: span.traceId,
      });
    });
  }

  await Promise.all([handle('req-A', 30), handle('req-B', 5)]);

  console.log('--- Two requests interleaved on one thread ---');
  for (const r of results) {
    const storeOk = r.withStore === r.correct ? 'ok' : 'WRONG';
    const globalOk = r.withGlobal === r.correct ? 'ok' : 'WRONG';
    console.log(`  ${r.request}  AsyncLocalStorage=${storeOk}   module-global=${globalOk}`);
  }
  console.log('  req-B finished first and overwrote the module-global, so req-A');
  console.log('  attributes its own work to req-B. Not a truncated trace: a');
  console.log('  complete, confident, wrong one.');
  console.log();
}

async function main() {
  console.log('Layer 6 Topic 3 - losing trace context in Node.js (AsyncLocalStorage)');
  console.log(`node ${process.version}   ${process.platform}/${process.arch}`);
  console.log('='.repeat(72));
  console.log();

  const verdicts = {};
  verdicts['await / timers'] = await boundaryAwait();

  const pool = new LeakyPool();
  verdicts['pool callback queue'] = await boundaryPool(pool);
  pool.close();

  verdicts['queue'] = boundaryQueue();
  verdicts['http traceparent'] = boundaryHttp();

  await interleaved();

  console.log('--- Summary: which boundaries Node covers for you ---');
  for (const [name, verdict] of Object.entries(verdicts)) {
    const who = verdict === 'preserved' ? 'runtime carries it' : 'YOU carry it';
    console.log(`  ${name.padEnd(22)} ${verdict.padEnd(10)} ${who}`);
  }
  console.log();

  console.log('--- The one-query test, on the log lines this run emitted ---');
  const withId = capturedLogs.filter((r) => r.trace_id).length;
  console.log(`  log lines emitted            ${capturedLogs.length}`);
  console.log(`  lines carrying a trace_id    ${withId}`);
  console.log(`  lines carrying nothing       ${capturedLogs.length - withId}   <- unqueryable by request`);
  for (const rec of capturedLogs) {
    console.log(`    ${rec.msg.padEnd(28)} trace_id=${rec.trace_id || '(empty)'}`);
  }
}

main();

File: 06-observability/03-correlation-ids/python/lose_the_context.py
"""
Layer 6 Topic 3 - Losing trace context at a Python concurrency boundary.

What this demonstrates
----------------------
The cross-process half of propagation is a wire format (`traceparent`) and it
is boring. The in-process half is the runtime's problem, and Python's answer is
`contextvars`:

  * asyncio understands contextvars. `asyncio.create_task` COPIES the current
    context into the new task, so the span you started is still current inside
    the task. This is the control run: it works, and it works silently, which
    is why the next one surprises people.
  * `loop.run_in_executor` does not. The callable runs on a thread-pool thread
    that was created before your request existed and never saw your context, so
    `CURRENT_SPAN.get()` there returns the default. The fix is to hand the
    executor `contextvars.copy_context().run` instead of the bare callable.
  * A queue is not a concurrency boundary the runtime can help with at all:
    there is no wire and no context, only a message body. If you do not put
    `traceparent` in the body yourself, nothing carries it.

Nothing here imports OpenTelemetry -- no SDK is installed on this machine. The
context variable, the span and the `traceparent` codec are ~40 lines of
standard library, which is the point: this failure is not a bug in an SDK, it
is a property of where the runtime does and does not copy context for you.

What to look for in the output
------------------------------
Four blocks, all in the same shape:

  caller trace_id   <id>
  callee trace_id   <id or "none">   naive
  callee trace_id   <id>             propagated
  verdict           lost | preserved

Read the verdict column top to bottom. `create_task` is preserved with no
effort; `run_in_executor` is lost with no warning; the queue is lost even
though both ends are in the same process. Then read the last section: the log
lines emitted from inside the executor thread carry no trace_id in the naive
run, which is exactly the "one-query test" failing.
"""
import asyncio
import contextvars
import json
import logging
import os
import queue
import random
import sys
import threading
from concurrent.futures import ThreadPoolExecutor

# ---------------------------------------------------------------------------
# A minimal span + W3C traceparent codec. This is the whole of the
# cross-process half of propagation, and it fits in a screen.
# ---------------------------------------------------------------------------

_RNG = random.Random(20260818)  # fixed seed: IDs differ per span, not per run


class Span:
    __slots__ = ("trace_id", "span_id", "name", "sampled")

    def __init__(self, name, trace_id=None, sampled=True):
        self.name = name
        self.trace_id = trace_id or "%032x" % _RNG.getrandbits(128)
        self.span_id = "%016x" % _RNG.getrandbits(64)
        self.sampled = sampled

    def traceparent(self):
        return "00-%s-%s-%02x" % (self.trace_id, self.span_id, 1 if self.sampled else 0)

    @staticmethod
    def from_traceparent(header, name):
        # version-traceid-spanid-flags, e.g.
        # 00-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-01
        version, trace_id, parent_id, flags = header.split("-")
        if version != "00" or len(trace_id) != 32 or len(parent_id) != 16:
            raise ValueError("malformed traceparent: %r" % header)
        return Span(name, trace_id=trace_id, sampled=bool(int(flags, 16) & 1))


# The context variable. Its default is what "no context" looks like: None.
CURRENT_SPAN: "contextvars.ContextVar[Span | None]" = contextvars.ContextVar(
    "current_span", default=None
)


def current_trace_id():
    span = CURRENT_SPAN.get()
    return span.trace_id if span else "none"


# ---------------------------------------------------------------------------
# Logging: a filter that reads the CURRENT context per record, and a JSON
# formatter. Both halves are required for the one-query test; a filter that
# reads the span once at import time is the classic broken version, and it is
# broken in a way that looks like it works (every line gets an id, the same id).
# ---------------------------------------------------------------------------

CAPTURED_LOGS = []


class TraceContextFilter(logging.Filter):
    def filter(self, record):
        span = CURRENT_SPAN.get()  # per record, not per process
        record.trace_id = span.trace_id if span else ""
        record.span_id = span.span_id if span else ""
        return True


class JsonFormatter(logging.Formatter):
    def format(self, record):
        payload = {
            "level": record.levelname,
            "msg": record.getMessage(),
            "thread": threading.current_thread().name,
            "trace_id": getattr(record, "trace_id", ""),
        }
        line = json.dumps(payload)
        CAPTURED_LOGS.append(payload)
        return line


def build_logger():
    logger = logging.getLogger("lose_the_context")
    logger.setLevel(logging.INFO)
    handler = logging.StreamHandler(open(os.devnull, "w"))
    handler.setFormatter(JsonFormatter())
    handler.addFilter(TraceContextFilter())
    logger.addHandler(handler)
    logger.propagate = False
    return logger


LOG = build_logger()


def report(boundary, caller, naive, propagated, note=""):
    verdict = "preserved" if naive == caller else "lost"
    print("boundary          %s" % boundary)
    print("caller trace_id   %s" % caller)
    print("callee trace_id   %-32s naive" % naive)
    print("callee trace_id   %-32s propagated" % propagated)
    print("verdict           %s%s" % (verdict, ("   (%s)" % note) if note else ""))
    print()
    return verdict


# ---------------------------------------------------------------------------
# Boundary 1: asyncio.create_task -- the control. Context is COPIED into the
# task at creation, so this one works and keeps working.
# ---------------------------------------------------------------------------

async def boundary_create_task():
    span = Span("GET /orders")
    CURRENT_SPAN.set(span)

    async def child():
        return current_trace_id()

    observed = await asyncio.create_task(child())
    return report(
        "asyncio.create_task",
        span.trace_id,
        observed,
        observed,
        note="asyncio copies the context at task creation; nothing to fix",
    )


# ---------------------------------------------------------------------------
# Boundary 2: run_in_executor -- the one that bites. The pool thread predates
# the request and has its own (empty) context.
# ---------------------------------------------------------------------------

def pricing_call_sync(label):
    """A synchronous downstream call, the reason you reached for an executor."""
    LOG.info("calling pricing (%s)", label)
    return current_trace_id()


async def boundary_run_in_executor(pool):
    span = Span("GET /orders")
    CURRENT_SPAN.set(span)
    loop = asyncio.get_running_loop()

    # Naive: the bare callable. The pool thread runs it in ITS context.
    naive = await loop.run_in_executor(pool, pricing_call_sync, "naive")

    # Propagated: copy_context() snapshots this task's context; ctx.run
    # installs it on the pool thread for the duration of the call.
    ctx = contextvars.copy_context()
    fixed = await loop.run_in_executor(pool, ctx.run, pricing_call_sync, "propagated")

    return report(
        "loop.run_in_executor",
        span.trace_id,
        naive,
        fixed,
        note="fix = run_in_executor(pool, contextvars.copy_context().run, fn, ...)",
    )


# ---------------------------------------------------------------------------
# Boundary 3: a queue. No wire, no runtime help. The message body is the only
# place a traceparent can live, and putting it there is entirely on you.
# ---------------------------------------------------------------------------

def worker_consume(job):
    """Runs on the worker thread. Only the message body crosses the boundary."""
    token = None
    if "traceparent" in job:
        token = CURRENT_SPAN.set(Span.from_traceparent(job["traceparent"], "job"))
    try:
        LOG.info("processing job %s", job["id"])
        return current_trace_id()
    finally:
        if token is not None:
            CURRENT_SPAN.reset(token)


def boundary_queue():
    span = Span("POST /orders")
    CURRENT_SPAN.set(span)

    q = queue.Queue()
    results = {}

    def worker():
        while True:
            job = q.get()
            if job is None:
                return
            results[job["id"]] = worker_consume(job)

    thread = threading.Thread(target=worker, name="worker-1", daemon=True)
    thread.start()

    q.put({"id": "naive", "customer": "cust-0042"})
    q.put({"id": "propagated", "customer": "cust-0042",
           "traceparent": span.traceparent()})
    q.put(None)
    thread.join()

    CURRENT_SPAN.set(span)  # the consumer ran on another thread; restore ours
    return report(
        "Postgres-backed queue",
        span.trace_id,
        results["naive"],
        results["propagated"],
        note="the transport carries no headers; put traceparent in the body",
    )


# ---------------------------------------------------------------------------
# Boundary 4: the outbound HTTP call. The easy half, shown so the wire format
# is concrete rather than a paragraph.
# ---------------------------------------------------------------------------

def boundary_http():
    span = Span("GET /orders")
    CURRENT_SPAN.set(span)
    header = span.traceparent()

    # ...over the wire to `pricing`, which parses it and continues the trace.
    downstream = Span.from_traceparent(header, "GET /price")

    print("boundary          HTTP request to pricing")
    print("caller trace_id   %s" % span.trace_id)
    print("traceparent sent  %s" % header)
    print("callee trace_id   %-32s parsed from the header" % downstream.trace_id)
    print("verdict           preserved   (this is what being a W3C standard buys)")
    print()
    return "preserved"


def main():
    print("Layer 6 Topic 3 - losing trace context in Python (contextvars)")
    print("python %s   %s" % (sys.version.split()[0], sys.platform))
    print("=" * 72)
    print()

    verdicts = {}
    pool = ThreadPoolExecutor(max_workers=2, thread_name_prefix="pool")
    try:
        # The pool threads are created here, before any request exists. That
        # single fact is the whole of boundary 2.
        list(pool.map(lambda _: None, range(2)))

        verdicts["asyncio.create_task"] = asyncio.run(boundary_create_task())
        verdicts["run_in_executor"] = asyncio.run(boundary_run_in_executor(pool))
    finally:
        pool.shutdown()

    verdicts["queue"] = boundary_queue()
    verdicts["http traceparent"] = boundary_http()

    print("--- Summary: which boundaries Python covers for you ---")
    for name, verdict in verdicts.items():
        covered = "runtime carries it" if verdict == "preserved" else "YOU carry it"
        print("  %-22s %-10s %s" % (name, verdict, covered))
    print()

    print("--- The one-query test, on the log lines this run emitted ---")
    with_id = [rec for rec in CAPTURED_LOGS if rec["trace_id"]]
    without = [rec for rec in CAPTURED_LOGS if not rec["trace_id"]]
    print("  log lines emitted            %d" % len(CAPTURED_LOGS))
    print("  lines carrying a trace_id    %d" % len(with_id))
    print("  lines carrying nothing       %d   <- unqueryable by request" % len(without))
    for rec in CAPTURED_LOGS:
        print("    %-9s %-24s trace_id=%s"
              % (rec["thread"].split("_")[0][:9], rec["msg"],
                 rec["trace_id"] or "(empty)"))
    print()
    print("  Every line above was emitted by the same logger with the same")
    print("  filter attached. The empty ones are not a logging bug: the filter")
    print("  read the context correctly and there was nothing in it. Fix the")
    print("  propagation and the logging fixes itself.")


if __name__ == "__main__":
    main()
